core/src/transaction_ffi.rs

```diff
@@ -0,0 +1,324 @@
+                num_accounts: instruction.accounts.len() as u16,
+                accounts: instruction.accounts.as_ptr(),
+                data_len: instruction.data.len() as u16,
+                data: instruction.data.as_ptr(),
+            };
+
+            if !unsafe { (callback.callback)(callback.state, instruction) } {
+                break;
+            }
+        }
+    }
+
+    extern "C" fn num_accounts<Tx: SVMTransaction>(transaction_ptr: TransactionPtr) -> usize {
+        let transaction = unsafe { &*(transaction_ptr as *const Tx) };
+        transaction.account_keys().len()
+    }
+
+    extern "C" fn get_account<Tx: SVMTransaction>(
+        transaction_ptr: TransactionPtr,
+        index: usize,
+    ) -> *const u8 {
+        let transaction = unsafe { &*(transaction_ptr as *const Tx) };
+        transaction
+            .account_keys()
+            .get(index)
+            .map_or(core::ptr::null(), |key| key.as_ref().as_ptr())
+    }
+
+    extern "C" fn is_writable<Tx: SVMTransaction>(
+        transaction_ptr: TransactionPtr,
+        index: usize,
+    ) -> bool {
+        let transaction = unsafe { &*(transaction_ptr as *const Tx) };
+        transaction.is_writable(index)
+    }
+
+    extern "C" fn is_signer<Tx: SVMTransaction>(
+        transaction_ptr: TransactionPtr,
+        index: usize,
+    ) -> bool {
+        let transaction = unsafe { &*(transaction_ptr as *const Tx) };
+        transaction.is_signer(index)
+    }
+
+    extern "C" fn is_invoked<Tx: SVMTransaction>(
+        transaction_ptr: TransactionPtr,
+        index: usize,
+    ) -> bool {
+        let transaction = unsafe { &*(transaction_ptr as *const Tx) };
+        transaction.is_invoked(index)
+    }
+
+    extern "C" fn iter_accounts<Tx: SVMTransaction>(
+        transaction_ptr: TransactionPtr,
+        callback: AccountCallback,
+    ) {
+        let transaction = unsafe { &*(transaction_ptr as *const Tx) };
+        for account_key in transaction.account_keys().iter() {
+            let account_key = account_key.as_ref().as_ptr();
+            if !unsafe { (callback.callback)(callback.state, account_key) } {
+                break;
+            }
+        }
+    }
+
+    TransactionInterface {
+        transaction_ptr: transaction as *const Tx as *const core::ffi::c_void,
+        num_signatures_fn: num_signatures::<Tx>,
+        signatures_fn: signatures::<Tx>,
+        num_total_signatures_fn: num_total_signatures::<Tx>,
+        num_write_locks_fn: num_write_locks::<Tx>,
+        recent_blockhash_fn: recent_blockhash::<Tx>,
+        num_instructions_fn: num_instructions::<Tx>,
+        iter_instructions_fn: iter_instructions::<Tx>,
+        num_accounts_fn: num_accounts::<Tx>,
+        get_account_fn: get_account::<Tx>,
+        is_writable_fn: is_writable::<Tx>,
+        is_signer_fn: is_signer::<Tx>,
+        is_invoked_fn: is_invoked::<Tx>,
+        iter_accounts_fn: iter_accounts::<Tx>,
+    }
```

Comment:
Discussed a similar question here: https://github.com/anza-xyz/agave/pull/3434#issuecomment-2469158222. But, wondering if it is better to just define `TransactionInterface` as:
```
#[repr(C)]
pub struct TransactionInterface {
    pub transaction_ptr: TransactionPtr,
}

// When creating the interface, you only set the pointer
pub unsafe fn create_transaction_interface<Tx: SVMTransaction>(
    transaction: &Tx,
) -> TransactionInterface {
    TransactionInterface {
        transaction as *const Tx as *const core::ffi::c_void,
    }
}
```
and then we define all of the `extern "C"` methods outside of `TransactionInterface`. That way we don't have to update the `TransactionInterface` if a new `Transaction` method is added and it won't break downstream consumers. Downside is the `extern "C"` methods would be in the global namespace. 

Curious to hear thoughts!

Reply:
Maybe I misunderstand the suggestion. How does a plugin know which functions to use then?

Not breaking downstream with current approach would just be passing a pointer to `TransactionInterface`. It's repr(c), so any new functions are just added on the end, it only breaks user-code if we remove/change things.

Reply:
Might not be what you have in mind, but I would actually like if we separated the interface into parts; something like

```rust
#[repr(C)]
pub struct TransactionInterface {
    pub transaction_ptr: TransactionPtr,
    pub transaction_function_table_ptr: TransactionFunctionTablePtr,
}
```

this way we can just use static table pointers to pass for the different transaction types we may have.

Reply:
ya, actually i take it back. what you're suggesting with the static table pointer makes more sense. and ya new functions just have to be added at the end of the `TransactionFunctionTablePtr` struct